### PR TITLE
Typing emoji

### LINKED ISSUE #559: Typing emoji
https://github.com/kblok/puppeteer/pull/10/commits/3335d369d3c5054517db7cdc541aba31184b9d07

### PR DESCRIPTION
**Pull request by kblok**
closes #559

### CODE REVIEW

File: lib/PuppeteerSharp.Tests/InputTests/InputTests.cs [hunk 1]
```diff
@@ -595,5 +595,28 @@ public async Task ShouldThrowOnUnknownKeys()
 
             await Assert.ThrowsAsync<KeyNotFoundException>(() => Page.Keyboard.PressAsync("😊"));
         }
+
+        [Fact]
+        public async Task ShouldTypeEmoji()
+        {
+            await Page.GoToAsync(TestConstants.ServerUrl + "/input/textarea.html");
+            await Page.TypeAsync("textarea", "👹 Tokyo street Japan \uD83C\uDDEF\uD83C\uDDF5");
```
Comment (kblok) [inline lib/PuppeteerSharp.Tests/InputTests/InputTests.cs]: I'm adding the Japanese flag as Unicode because it breaks the IDE :joy:

File: lib/PuppeteerSharp/Input/Keyboard.cs [hunk 2]
```diff
@@ -109,8 +103,11 @@ public async Task TypeAsync(string text, TypeOptions options = null)
             {
                 delay = (int)options.Delay;
             }
-            foreach (var letter in text)
+
+            var textParts = StringInfo.GetTextElementEnumerator(text);
```
Comment (Meir017) [inline lib/PuppeteerSharp/Input/Keyboard.cs]: why was this change required?

File: lib/PuppeteerSharp/Input/Keyboard.cs [hunk 3]
```diff
@@ -41,7 +42,7 @@ public Task DownAsync(string key, DownOptions options = null)
 
             var text = options?.Text == null ? description.Text : options.Text;
 
-            return _client.SendAsync("Input.dispatchKeyEvent", new Dictionary<string, object>(){
+            return _client.SendAsync("Input.dispatchKeyEvent", new Dictionary<string, object>{
```
Comment (Meir017) [inline lib/PuppeteerSharp/Input/Keyboard.cs]: formatting ;)

Comment (kblok) [reply lib/PuppeteerSharp/Input/Keyboard.cs]: Done

File: lib/PuppeteerSharp/Input/Keyboard.cs (hunk 2 above)
Comment (kblok) [reply lib/PuppeteerSharp/Input/Keyboard.cs]: I was just preparing a post about this. Emojis (let's call it Unicode chars) uses 4 bytes. So you need 2 chars to represent one Unicode char. If you do `string.ToCharArray` and send one by one to Chromium you would get invalid characters (a.k.a. □) on the other side.
So you need a method, `GetTextElementEnumerator`, that returns the "elements" of the text. `GetTextElementEnumerator` is able to return a \u1F479 (which are 2 chars: \u1F47 and 9) as one element.
From the [docs](https://docs.microsoft.com/en-us/dotnet/api/system.globalization.stringinfo.gettextelementenumerator?view=netframework-4.7.2):
>The .NET Framework defines a text element as a unit of text that is displayed as a single character, that is, a grapheme. A text element can be a base character, a surrogate pair, or a combining character sequence.
